tests/nextflowtower/test_client.py

```diff
@@ -0,0 +1,111 @@
+import json
+import os
+from types import GeneratorType
+
+import pytest
+
+from sagetasks.nextflowtower import client
+
+EG_TOKEN = "token"
+EG_API_URL = "https://example.com"
+EG_METHOD = "GET"
+EG_ENDPOINT = "/example/endpoint"
+
+
+@pytest.fixture
+def tower_client():
+    return client.TowerClient(EG_TOKEN, EG_API_URL)
+
+
+def test_import():
+    assert client
+
+
+class TestTowerClient:
+    def test_missing_args(self):
+        # Clear environment
+        env_vars = [
+            "NXF_TOWER_TOKEN",
+            "TOWER_ACCESS_TOKEN",
+            "NXF_TOWER_API_URL",
+            "TOWER_API_ENDPOINT",
+        ]
+        for var in env_vars:
+            if var in os.environ:
+                del os.environ[var]
+        # Ensure exception when initializing
+        with pytest.raises(KeyError):
+            client.TowerClient()
+
+    def test_get_valid_name(self, tower_client):
+        expectations = {
+            "foobar": "foobar",
+            "FooBar": "FooBar",
+            "foo-bar": "foo-bar",
+            "foo_bar": "foo_bar",
+            "f00b4r": "f00b4r",
+            "foo,bar": "foo-bar",
+            "foo (bar)": "foo--bar-",
+        }
+        for name, output in expectations.items():
+            assert tower_client.get_valid_name(name) == output
+
+    def test_request(self, mocker, capfd, tower_client):
+        # Setup
+        eg_kwargs = {"foo": "bar"}
+        mocked_request = mocker.patch("requests.request", autospec=True)
+        mocked_request.return_value.json.return_value = eg_kwargs
+
+        def raise_json_error():
+            raise json.decoder.JSONDecodeError("", "", 0)
+
+        # Regular call with non-empty response (w/o debugging)
+        result = tower_client.request(EG_METHOD, EG_ENDPOINT, params=eg_kwargs)
+        (method, url), kwargs = mocked_request.call_args
+        captured = capfd.readouterr()
+        mocked_request.assert_called_once()
+        assert method == EG_METHOD
+        assert url.startswith(EG_API_URL)
+        assert url.endswith(EG_ENDPOINT)
+        assert "Authorization" in kwargs["headers"]
+        assert kwargs["headers"]["Authorization"] == f"Bearer {EG_TOKEN}"
+        assert "params" in kwargs
+        assert kwargs["params"] == eg_kwargs
+        assert result == eg_kwargs
+        assert captured.out == ""
+
+        # Regular call with empty response (w/ debugging)
```

Comment:
Nit: Personally, I would split these up into two test functions, but I think this is good.

Reply:
Agreed!